lib/sensors/GPS.cpp

```diff
@@ -0,0 +1,476 @@
+	//TODO: need error checking for failed checksums, etc
+	while (!m_ack_flag) { //wait for ack
+		if (m_nack_flag || m_ack_timeout.read_ms() > ACK_TIMEOUT_MS) {
+			m_ack_timeout.stop();
+			m_target_class = tmpcls;
+			m_target_msgID = tmpID;
+			return false;
+		}
+	}
+	m_target_class = tmpcls;
+	m_target_msgID = tmpID;
+	return true;
+}
+
+void GPS::onByteReceive() {
+	process(m_uart.getc(), (m_activePacketBuffer == UBLOX_PACKET_PACKETCFG ? &m_packetCfg :
+							m_activePacketBuffer == UBLOX_PACKET_PACKETACK ? &m_packetAck : &m_packetBuf),
+			m_target_class, m_target_msgID);
+}
+
+void GPS::sendSerialCommand(ubxPacket *outgoingUBX) {
+	calcChecksum(outgoingUBX);
+	//Write header bytes
+	m_uart.putc(UBX_SYNCH_1); //μ
+	m_uart.putc(UBX_SYNCH_2); //b
+	m_uart.putc(outgoingUBX->cls);
+	m_uart.putc(outgoingUBX->id);
+	m_uart.putc(outgoingUBX->len & 0xFF); //LSB
+	m_uart.putc(outgoingUBX->len >> 8);   //MSB
+
+	//Write payload
+	m_uart.write(outgoingUBX->payload, outgoingUBX->len, NULL);
+
+	//Write checksum
+	m_uart.putc(outgoingUBX->checksumA);
+	m_uart.putc(outgoingUBX->checksumB);
+}
+
+//this whole processing chain is a mess, I did my best to port it from the sparkfun library
+//ideally a total refactor can be done at some point
+//lots of stuff in here we wont use, and lots that seems VERY fragile
+//I have my doubts that this works at all honestly, but I can't say for sure until I test it
+void GPS::process(uint8_t incoming, ubxPacket *incomingUBX, uint8_t requestedClass, uint8_t requestedID) {
+	if (currentSentence == NONE) {
+		if (incoming == 0xB5) //UBX binary frames start with 0xB5, aka μ
+		{
+			//start of sentence
+			m_ubxFrameCounter = 0;
+			currentSentence = UBX;
+			m_packetBuf.counter = 0;
+			m_ignoreThisPayload = false;
+			//Store data in m_packetBuf until class known
+			m_activePacketBuffer = UBLOX_PACKET_PACKETBUF;
+		} else {
+			//TODO: handle failure case (missed sentence start)
+		}
+	}
+
+	//non-payload bytes if/else chain
+	if (((m_ubxFrameCounter == 0) && (incoming != 0xB5)) ||     //'μ'
+			((m_ubxFrameCounter == 1) && (incoming != 0x62)))  //'b'
+		currentSentence = NONE;        //error, reset sentence
+	else if (m_ubxFrameCounter == 2) //Class
+	{
+		m_rollingChecksumA = 0; //reset member states
+		m_rollingChecksumB = 0;
+		m_packetBuf.counter = 0;
+		m_packetBuf.valid = UBLOX_PACKET_VALIDITY_NOT_DEFINED;
+		m_packetBuf.startingSpot = incomingUBX->startingSpot;
+	} else if (m_ubxFrameCounter == 3) { //ID
+		//determine class
+		if (m_packetBuf.cls != UBX_CLASS_ACK) {
+			if ((m_packetBuf.cls == requestedClass) && (m_packetBuf.id == requestedID)) {
+				//not ACK, Class and ID match
+				//copy data to proper buffer
+				m_activePacketBuffer = UBLOX_PACKET_PACKETCFG;
+				incomingUBX->cls = m_packetBuf.cls;
+				incomingUBX->id = m_packetBuf.id;
+				incomingUBX->counter = m_packetBuf.counter;
+			} else {
+				//not ACK or requested packet, so ignore it
+				m_ignoreThisPayload = true;
+			}
+		}
```

Comment:
The sparkfun library implementation stores the lengths of the LSB and MSB when `ubxFrameCounter` is at 4 and 5, was that deliberately taken out here?